### ISSUE TITLE
schema: Reproducible image hashes

### ISSUE THREAD

**Issue by petrosagg** [appc/spec#199]
Currently an image hash is computed by creating a tar archive of `manifest` and `rootfs` and taking the hash of it. By default, tar includes files in the archive in the order they are returned by `readdir()`, which can differ between hosts/filesystems. By enforcing an ordering of the files in the tar archive, computing an image ID of an unpacked image will always give the same result.

## Why is this useful?
This becomes very useful when someone wants to combine signed images with some sort of delta transport. What I mean by that is a server (possibly untrusted) that can compute and serve an image delta from image A to image B.

With that, a client that already has image A can get and verify image B by:
* Downloading A->B delta file
* Downloading detached image B signature
* Apply the delta on a copy of image A
* Run the reproducible tar and pipe it to `gpg --verify`

## How can this be done?
New versions of tar support a `--sort=name` option that does exactly that. The same thing can be accomplished in with an old tar by creating a sorted list of files and piping the list to tar with the `--files-from=-` option.

**Comment by petrosagg** [comment]
To give some context, I'm the tech lead for [resin.io](https://resin.io/) where we ship containers to embedded devices. In our use case bandwidth usually is not plenty and certainly not cheap so an efficient **and** verifiable transport makes a lot of sense for us.

**Comment by jonboulle** [comment]
Seems reasonable to me. @philips @cdaylward ?

(Relates to #146, #186, will be superseded by streaming image format but that's a longer-term plan)

**Comment by petrosagg** [comment]
What are the requirements for the streaming image format? Is there any WIP spec for that?

**Comment by philips** [comment]
@petrosagg I agree that being able to make deltas and apply them as a bandwidth saving measure is desirable! In fact early versions of the spec had the manifest holding a hash of the contents of every file and the path/perm/etc of those objects but it overly-complicated the spec.

I am curious at why the file ordering matters for the diff server you are suggesting.

**Comment by petrosagg** [comment]
@philips the file ordering is irrelevant to the diffing server. It on the client-side after having applied the diff that it is needed. Basically, file ordering lets you recreate locally the bytestream that you would have gotten if you downloaded the whole tar archive, and therefore verify it. I hope this makes sense.

**Comment by petrosagg** [comment]
As a matter of fact, the diffing server could compute diffs at the block level, without even caring about the files. For example using btrfs send/receive.

**Comment by philips** [comment]
I am not quite following. Could you give some psuedo code or a link to a similar system?

**Comment by petrosagg** [comment]
Ok, let's say Alice creates two images, `a.aci` and `b.aci` and then creates two signatures, `a.aci.asc` and `b.aci.asc` with her private key. She then proceeds and uploads these files to a third party server.

Bob has already downloaded, verified and unpacked `a.aci`. This means that his disk looks like this:

```bash
/a/manifest
/a/rootfs
/a/rootfs/usr
...
```

To efficiently get `b.aci` he proceeds to do the following:
```bash
# Create a copy of image A
cp -r /a /b
# Use rsync to get the *unpacked* image B without wasting bandwidth
rsync -avz remote.example.org:/b /b
# Downloads the detached signature for b.aci
wget remote.example.org/b.aci.asc
```

Now Bob's `/b/` directory contains the contents of image B, but of course he wants to verify that. The only signature he has at his disposal is `b.aci.asc` and as such, to verify that his unpacked image is indeed image B he has to recreate `b.aci` **in the same exact way Alice did it when she created the image** and then use gpg to verify it.

This is where a simple `tar` will fail. Simply issuing `tar cfv b.aci manifest rootfs` is not guaranteed to create the exact same archive across hosts/filesystems and therefore the signature won't be valid.

Copying from the tar man page:
```
Using --sort=name ensures the member ordering in the created archive is uniform and reproducible.
```

**Comment by jonboulle** [comment]
Yep, this doesn't get us all the way to perfect reproducibility but is
another good incremental step
Am 16.02.2015 17:11 schrieb "Petros Aggelatos" <notifications@github.com>:

**Comment by petrosagg** [comment]
Is there a particular edge case you have in mind?

**Comment by jonboulle** [comment]
Not in particular with ordering, just more generally with tar (i.e. that different tools can produce slightly different tarballs)

**Comment by philips** [comment]
I have retired myself to the fact that generating a 100% correct tarball everytime from the same source filesystem won't work. See https://github.com/coreos/rocket/issues/544

In any case I am OK enforcing this constraint of the tar files in the ACIs. So, who wants to update the spec and write the code? :)

**Comment by jonboulle** [comment]
Well that is an implementation issue with golang in particular, right?

**Comment by petrosagg** [comment]
@philips I'll build upon Tor's [deterministic tar script](https://gitweb.torproject.org/builders/tor-browser-bundle.git/tree/gitian/build-helpers/dtar.sh?id=681b1598bbbb0704f620bc0e27624a46d6f51f89) and also use some the [recommendations](https://wiki.debian.org/ReproducibleBuilds/Howto#Identified_problems.2C_and_possible_solutions) that make sense for this usecase from Debian's reproducible builds and send a PR.

**Comment by philips** [comment]
@petrosagg OK, I think the place to start is to:

- [ ] patch SPEC.md to say that tar files must be sorted
- [ ] patch to `actool validate` to ensure that an ACI is in sorted order
- [ ] add tests to `actool build` to ensure it does in order walking. This is the default for Go though "filepath.Walk - The files are walked in lexical order, which makes the output deterministic but means that for very large directories Walk can be inefficient. Walk does not follow symbolic links."

**Comment by jonboulle** [comment]
https://github.com/docker/docker/pull/14067 is an example of using tar-split to recreate the tar files in a deterministic way
